docs/adr/0003-consensus-runtime-token-transfer.md

```diff
@@ -0,0 +1,379 @@
+runtime thorugh a `Withdraw` message._
+
+### Messages
+
+Since this is the first proposal that introduces a new runtime message type that
+can be emitted from a runtime during a round, it also defines some general
+properties of runtime messages and the dispatch mechanism:
+
+- Each message has an associated gas cost that needs to be paid by the
+  submitter (e.g. as part of the `roothash.ExecutorCommit` method call).
+
+- There is a maximum number of messages that can be emitted by a runtime during
+  a given round. The limit is defined both globally (e.g. a roothash consensus
+  parameter) and per-runtime (which needs to be equal to or lower than the
+  global limit).
+
+- Messages are serialized using a sum type describing all possible messages,
+  where each message type is assigned a _field name_:
+
+  ```golang
+  type Message struct {
+      Message1 *Message1 `json:"message1,omitempty"`
+      Message2 *Message2 `json:"message2,omitempty"`
+      // ...
+  }
+  ```
+
+- A change is made to how messages are included in commitments, to reduce the
+  size of submitted transactions.
+
+  The `ComputeResultsHeader` is changed so that the `Messages` field is replaced
+  with a `MessagesHash` field containing a hash of the CBOR-encoded messages
+  emitted by the runtime.
+
+  At the same time `ComputeBody` is changed to include an additional field
+  `Messages` as follows:
+
+  ```golang
+  type ComputeBody struct {
+      // ... existing fields omitted ...
+      Messages []*block.Message `json:"messages,omitempty"`
+  }
+  ```
+
+  The `Messages` field must only be populated in the commitment by the
+  transaction scheduler and must match the `MessagesHash`.
+
+- If any of the included messages is deemed _invalid_, the round fails and the
+  runtime state is not updated.
+
+This proposal introduces the following runtime messages:
+
+#### Transfer
+
+The transfer message enables transfers between runtime sub accounts or from
+runtime sub account to runtime account.
+
+**Field name:**
+
+```
+transfer
+```
+
+**Body:**
+
+```golang
+type TransferMessage struct {
+    From   *staking.Address  `json:"from,omitempty"`
+    To     *staking.Address  `json:"to,omitempty"`
+    Amount quantity.Quantity `json:"amount"`
+}
+```
+
+**Fields:**
+
+- `from` specifies the source account (emitting runtime sub account is implied).
+  If not specified, the source account is the runtime account.
+- `to` specifies the destination account (emitting runtime sub account is
+  implied). If not specified, the destination account is the runtime account.
+- `amount` specifies the amount of base units to transfer.
+
+If neither `from` nor `to` are specified the message is invalid.
+
+If there is not enough balance in the given source account (either the runtime
+sub account or the runtime account), the message is invalid.
+
+#### Withdraw
+
+The withdraw message enables transfers from a runtime sub account to a general
```

Comment:
could the name be "refund", "release", or something like that instead of "withdraw".  the actor for an withdrawal action is usually the depositor, and i think it might be confusing since (i think of) the depositor is the owner of the general account, and this message is only sent by the runtime.

Reply:
Yeah good point, "refund" sounds good.

Reply:
This has been refactored to the allowance model (where an account holder can approve beneficiaries so they can withdraw) where I think withdraw makes more sense. Please take another look.